prince/interaction_rates.py

```diff
@@ -250,95 +262,63 @@ def _init_matrices(self):
 
                 # Finally write this to the batch matrix
                 self._batch_matrix[ibatch:ibatch + len(emoidx), :] = res
-                self._batch_rows.append(edaidx[:,0])
-                self._batch_cols.append(emoidx[:,0])
+                self._batch_rows.append(edaidx[:, 0])
+                self._batch_cols.append(emoidx[:, 0])
                 ibatch += len(emoidx)
 
             else:
                 info(20, 'Species combination not included in model', moid,
                      daid)
 
-        #self._batch_matrix.resize((ibatch, dph))
-        self._batch_matrix = self._batch_matrix[:ibatch,:]
-        # self._batch_rows = np.array(self._batch_rows)
-        # self._batch_cols = np.array(self._batch_cols)
+        self._batch_matrix = self._batch_matrix[:ibatch, :]
         self._batch_rows = np.concatenate(self._batch_rows, axis=None)
         self._batch_cols = np.concatenate(self._batch_cols, axis=None)
         self._batch_vec = np.zeros(ibatch)
 
-        info(2,f'Batch matrix shape: {self._batch_matrix.shape}')
-        info(2,f'Batch rows shape: {self._batch_rows.shape}')
-        info(2,f'Batch cols shape: {self._batch_cols.shape}')
-        info(2,f'Batch vector shape: {self._batch_vec.shape}')
+        info(2, f'Batch matrix shape: {self._batch_matrix.shape}')
+        info(2, f'Batch rows shape: {self._batch_rows.shape}')
+        info(2, f'Batch cols shape: {self._batch_cols.shape}')
+        info(2, f'Batch vector shape: {self._batch_vec.shape}')
 
         memory = (self._batch_matrix.nbytes + self._batch_rows.nbytes +
                   self._batch_cols.nbytes + self._batch_vec.nbytes) / 1024**2
         info(3, "Memory usage after initialization: {:} MB".format(memory))
 
-    def _init_coupling_mat(self, sp_format):
-        """Initialises the coupling matrix directly in sparse (csc) format.
+    def _init_coupling_mat(self):
+        """Initialises the coupling matrix directly in sparse (csr) format.
         """
-        info(2, 'Initiating coupling matrix in ({:}) format'.format(sp_format))
-
-        if sp_format == 'csc':
```

Comment:
The reason is that the MKL code path expects CSR and would require calling explicitly mkl_cscmv/mm instead of mkl_csrmv/mm in the ctypes interface. The same is valid for cupy's sparse matrix constructors. 
This adds some additional complexity and optional code path that needs to be maintained and actually tested.

I think we should re-check at some point the matrix storage. If it was faster then we can switch to csc entirely and consistently, but keeping it as an option is bad style.

In future it may be worth to decompose the matrix in the dense parts (block u-triangular) for the redistributed parts and the block diagonal, and use a mixture of dense and sparse algebra in a solver.